lib/ex_unit/lib/ex_unit/runner.ex

```diff
@@ -157,35 +179,82 @@ defmodule ExUnit.Runner do
     parent_pid = self()
     timeout = get_timeout(config, %{})
 
-    {module_pid, module_ref} =
-      spawn_monitor(fn ->
-        ExUnit.OnExitHandler.register(self())
-
-        case exec_module_setup(test_module) do
-          {:ok, test_module, context} ->
-            finished_tests = Enum.map(tests, &run_test(config, &1, context))
-            send(parent_pid, {self(), :module_finished, test_module, [], finished_tests})
-
-          {:error, test_module} ->
-            failed_tests = Enum.map(tests, &%{&1 | state: {:invalid, test_module}})
-            send(parent_pid, {self(), :module_finished, test_module, failed_tests, []})
-        end
-
-        exit(:shutdown)
-      end)
+    {module_pid, module_ref} = spawn_module_monitor(config, test_module, parent_pid, tests)
 
-    {test_module, pending, finished_tests} =
+    {test_module, test_results} =
       receive do
-        {^module_pid, :module_finished, test_module, failed_tests, finished_tests} ->
-          Process.demonitor(module_ref, [:flush])
-          {test_module, failed_tests, finished_tests}
+        {^module_pid, :module_finished, test_module, tests} ->
+          if max_failures_reached?(config.stats_pid, config.max_failures) do
+            Process.demonitor(module_ref, [:flush])
+            {test_module, %{pending: [], finished: []}}
```

Comment:
Let's continue with the three element tuple here too. No need to put it in a map only to extract it out.